### PR TITLE
[Feat] Add JSSP environment

### PR DESCRIPTION
**Pull request by LTluttmann**
## Description

- Added environment for the Job-Shop Scheduling Problem (JSSP). This implementation of JSSP treats JSSP as a special case of FJSP, where each operation can be processed by only one machine. As such, the environment is implemented as subclass of the FJSPEnv, only changing the action space (action space reduces to the selection of next job to execute) and the data generator. 
- In addition, the HetGNN policy has been restructured and renamed to L2D (for learning to dispatch) and is now applicable to both, FJSP and JSSP. 

## Motivation and Context

- JSSP is a common CO problem and a widely used problem to benchmark new algorithms

- [ ] I have raised an issue to propose this change ([required](https://github.com/ai4co/rl4co/issues) for new features and bug fixes)

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [ ] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: rl4co/models/__init__.py [hunk 1]
```diff
@@ -19,7 +19,7 @@
 from rl4co.models.rl.ppo.ppo import PPO
 from rl4co.models.rl.reinforce.baselines import REINFORCEBaseline, get_reinforce_baseline
 from rl4co.models.rl.reinforce.reinforce import REINFORCE
-from rl4co.models.zoo import HetGNNModel
+from rl4co.models.zoo import L2DModel
```
Comment (fedebotu) [inline rl4co/models/__init__.py:22]: Good - let's make sure the baselines have their names! L2D is a very influential paper in the NCO community

File: rl4co/utils/ops.py [hunk 2]
```diff
@@ -73,7 +73,8 @@ def gather_by_index(src, idx, dim=1, squeeze=True):
     expanded_shape = list(src.shape)
     expanded_shape[dim] = -1
     idx = idx.view(idx.shape + (1,) * (src.dim() - idx.dim())).expand(expanded_shape)
-    return src.gather(dim, idx).squeeze() if squeeze else src.gather(dim, idx)
+    squeeze = idx.size(dim) == 1 and squeeze
+    return src.gather(dim, idx).squeeze(dim) if squeeze else src.gather(dim, idx)
 
 
 def unbatchify_and_gather(x: Tensor, idx: Tensor, n: int):
```
Comment (fedebotu) [inline rl4co/utils/ops.py:80]: [Minor] it might be faster if you put a `@torch.jit` decorator. Not 100% sure though

**Review by fedebotu — APPROVED** [review]

**Comment by fedebotu** [comment]
@Junyoungpark tagging you since you know this problem very well!

Do you think we have a chance to include [ScheduleNet](https://arxiv.org/abs/2106.03051)?

File: rl4co/envs/scheduling/jssp/env.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+from tensordict import TensorDict
+from torch._tensor import Tensor
+
+from rl4co.envs import FJSPEnv
+from rl4co.utils.ops import gather_by_index
+
+from .generator import JSSPFileGenerator, JSSPGenerator
+
+
+class JSSPEnv(FJSPEnv):
```
[the remark this answers is not in the document]
Comment (cbhua) [reply rl4co/envs/scheduling/jssp/env.py:13]: Yesss, when see this inherit: wow nice.

File: rl4co/models/nn/graph/gcn.py [hunk 4]
```diff
@@ -15,7 +15,104 @@
...
 
+EdgeIndexFnSignature = Callable[[TensorDict, int, bool], Tensor]
+
+
+def edge_idx_fn_wrapper(td: TensorDict, num_nodes: int):
+    # self-loop is added by GCNConv layer
+    return get_full_graph_edge_index(td.device, num_nodes, self_loop=False)
+
+
 class GCNEncoder(nn.Module):
```
Comment (cbhua) [inline rl4co/models/nn/graph/gcn.py:26]: I like this clean refactoring, the logic is clearer. But will the `get_full_graph_edge_index()` be called at every forward step? i.e. in the previous version, if it's a fully connected connected graph, the `edge_index` will be saved as class variable, instead of regenerating every time.

**Review by cbhua — APPROVED** [review]
Great job!

File: rl4co/envs/scheduling/fjsp/env.py [hunk 5]
```diff
@@ -285,6 +293,17 @@ def _make_step(self, td: TensorDict, selected_job, selected_machine) -> TensorDi
         td["ma_assignment"][batch_idx, selected_machine, selected_op] = 1
         # update the state of the selected machine
         td["busy_until"][batch_idx, selected_machine] = td["time"] + proc_time_of_action
+        # update adjacency matrices (remove edges)
+        td["proc_times"] = td["proc_times"].scatter(
+            2,
+            selected_op[:, None, None].expand(-1, self.num_mas, 1),
```
[the remark this answers is not in the document]
Comment (fedebotu) [reply rl4co/envs/scheduling/fjsp/env.py:299]: Is it though? I think if you know already the dimensions, `einops` is slightly slower from what I know, but take this with a grain of salt. But I agree that it's more readable

[the remark this answers is not in the document]
Comment (cbhua) [reply rl4co/envs/scheduling/fjsp/env.py:299]: Okay I did a trial and actually `einops.repeat` is way slower than `tensor.expand` 😂 (in large scale around 4x slower). Then I think it's good to keep using the `tensor.expand`.

[the remark this answers is not in the document]
Comment (LTluttmann) [reply rl4co/envs/scheduling/fjsp/env.py:299]: yeah I think they use very clever optimizations in torch.expand() 😄

File: rl4co/models/nn/graph/gcn.py (hunk 4 above)
Comment (LTluttmann) [reply rl4co/models/nn/graph/gcn.py:26]: thats true, but the result is cached so it should not be too slow. In fact, this implementation should be much faster as before (at least it was in my experiments), because it avoids the list comprehension over the batch data within the forward pass. But I agree, its still not optimal; I will revisit this in the near future

**Comment by LTluttmann** [comment]
Thanks for reviewing guys. I will add a ton of changes here in a couple of minutes and I hope this PR will not get too messy. Let me know if we should go through it together.

Additional changes:
- stepwise PPO for L2D 
- Some MatNet changes to better work for JSSP / FJSP
- Running mean / variance class for reward / advantage scaling
- stepwise L2D policy
- attention based models for fjsp / jssp
- minor bugfixes and improvements here and there

**Comment by fedebotu** [comment]
Wow, lots of changes here! 😁 Really curious about episodic / stepwise RL performances


Btw, feel free to merge anytime

File: rl4co/envs/routing/tsp/env.py [hunk 6]
```diff
@@ -167,5 +167,59 @@ def check_solution_validity(td: TensorDict, actions: torch.Tensor):
...
     @staticmethod
-    def render(td: TensorDict, actions: torch.Tensor=None, ax = None):
+    def render(td: TensorDict, actions: torch.Tensor = None, ax=None):
         return render(td, actions, ax)
+
+
+# NOTE Experimental TSP class for stepwise PPO
+
+
+class TSPEnv4PPO(TSPEnv):
```
Comment (fedebotu) [inline rl4co/envs/routing/tsp/env.py:177]: [Minor] this may be called `DenseRewardTSPEnv` or similar?

File: rl4co/envs/scheduling/ffsp/env.py [hunk 7]
```diff
@@ -58,7 +57,7 @@ def __init__(
         generator_params: dict = {},
         **kwargs,
     ):
-        super().__init__(**kwargs)
+        super().__init__(check_solution=False, **kwargs)
```
Comment (fedebotu) [inline rl4co/envs/scheduling/ffsp/env.py:60]: Is this always the case (no solution check)?

**Review by fedebotu — APPROVED** [review]

File: rl4co/envs/routing/tsp/env.py (hunk 6 above)
Comment (LTluttmann) [reply rl4co/envs/routing/tsp/env.py:177]: Sure! Btw., Stepwise PPO for TSP indeed converges to the nearest neighbor heuristic, at least with the stepwise reward as it is defined here (the distance added by the action):

image

Do you have a preferrence as to how to call the stepwise PPO in the paper (dense, stepwise something else)? And then we should probably adjust the description about PPO in the appendix

File: rl4co/envs/scheduling/ffsp/env.py (hunk 7 above)
Comment (LTluttmann) [reply rl4co/envs/scheduling/ffsp/env.py:60]: uhm I think this is there just bc there is no check implemented for FFSP yet haha. Let me see if I can get one implemented

Comment (fedebotu) [reply rl4co/envs/scheduling/ffsp/env.py:60]: No worries, this is not a pressing issue! Actually, it's even better to keep it to `False` during training for efficiency

**Comment by fedebotu** [comment]
Great job!! This PR is truly huge
